fluentd-forward/teleport_client.go

```diff
@@ -0,0 +1,232 @@
+/*
+Copyright 2015-2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"time"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/types/events"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+// TeleportSearchEventsClient is an interface for client.Client, required for testing
+type TeleportSearchEventsClient interface {
+	SearchEvents(ctx context.Context, fromUTC, toUTC time.Time, namespace string, eventTypes []string, limit int, startKey string) ([]events.AuditEvent, string, error)
+	Close() error
+}
+
+// TeleportClient represents wrapper around Teleport client to work with events
+type TeleportClient struct {
+	// client is an instance of GRPC Teleport client
+	client TeleportSearchEventsClient
+
+	// cursor current page cursor value
+	cursor string
+
+	// nextCursor next page cursor value
+	nextCursor string
+
+	// id latest event returned by Next()
+	id string
+
+	// pos current virtual cursor position within a batch
+	pos int
+
+	// batch current events batch
+	batch []events.AuditEvent
+
+	// batchSize is fetch batch size
+	batchSize int
+
+	// namespace is events namespace
+	namespace string
+
+	// types is events types list
+	types []string
+
+	// startTime is event time frame start
+	startTime time.Time
+}
+
+// NewTeleportClient builds Teleport client instance
+func NewTeleportClient(c *Config, cursor string, id string) (*TeleportClient, error) {
+	var cl *client.Client
+	var err error
+
+	if c.TeleportIdentityFile != "" {
+		cl, err = newUsingIdentityFile(c)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		cl, err = newUsingKeys(c)
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
You don't need to switch on identity/keypair credentials, just provide both and `client.New` will use whichever one successfully connects.
```suggestion
	config := client.Config{
		Addrs: []string{c.TeleportAddr},
		Credentials: []client.Credentials{
			client.LoadKeyPair(c.TeleportCert, c.TeleportKey, c.TeleportCA),
			client.LoadIdentityFile(c.TeleportIdentityFile),
		},
	}
	if client, err = client.New(context.Background(), config); err != nil {
		return nil, err
	}
```

Reply:
Great, good to know, will change this piece.